Lists and tables design guideline update

doruskova opened the pull request:
I updated the design guidelines for Lists and tables to reflect the new preference for compact spacing. Includes: 
- updated images 
- section "Table/List capabilities"
- section "When to use compact vs. basic spacing "
- links to colors, bulk selection, compact and basic spacing have been added

Fix #1745

patternfly-build commented:
PF3 preview: https://patternfly-org-pr-1898-v3.surge.sh/v3

mmenestr commented on packages/v4/src/content/design-guidelines/usage-and-behavior/lists-and-tables/lists-and-tables.md:
```diff
@@ -8,20 +8,33 @@ showTitle: true
+* You want column headers.
 
 ## Data lists
 PatternFly supports several variations of the data list component.
```
I would add here "PatterFly supports two main types of data lists
1. Compact data list
2. Basic data list"

and then link both of those. Just to introduce both those types from the beginning, so that when you mention them again under "Data list capabilities" people know what you're talking about!

mmenestr commented on packages/v4/src/content/design-guidelines/usage-and-behavior/lists-and-tables/lists-and-tables.md:
```diff
@@ -60,7 +66,7 @@ Do not use actionable data lists when actions are restricted to a single row or
 The expandable list adds an expansion panel to every row to reveal more details about the item.
 
-![expandable data list](./img/expandable-list.png)
+<img src="./img/expandable-list.png"  alt="Expandable data list"  width="1174"/> 
```
Would get rid of the "bulk selects" in the expanded table - cause I'm not sure how those would interact with the main bulk select haha

mcoker replied:
<img width="772" alt="Screen Shot 2020-06-22 at 1 48 52 PM" src="https://user-images.githubusercontent.com/35148959/85324303-37571b80-b48f-11ea-91a8-d9bd3d383873.png">

the "2nd" marker should be on the expansion - looks like it's on the toggle.

mmenestr commented on packages/v4/src/content/design-guidelines/usage-and-behavior/lists-and-tables/lists-and-tables.md:
```diff
@@ -70,41 +76,46 @@ The expandable list adds an expansion panel to every row to reveal more details
 Use an expandable data list when you have more information than will comfortably fit inside a row, or you want to provide a way for advanced users to access information that is not applicable to all users.
 
 ## Tables
-* **Standard table:** Combines the styling of a list view with behaviors that are commonly associated with a table (sometimes referred to as a hybrid list)
```
How come you deleted the standard table and compact table info! I think it's useful to have it!
Would at a minimum have a similar description that you had in the data list section, so something simple like:
"PatterFly supports two main types of table"
- Compact  tables
- Standard tables"

mmenestr commented on packages/v4/src/content/design-guidelines/usage-and-behavior/lists-and-tables/lists-and-tables.md:
```diff
@@ -70,41 +76,46 @@ The expandable list adds an expansion panel to every row to reveal more details
+### Table capabilities
+Every [compact](#when-to-use-compact-vs-basic-spacing) or [basic](#when-to-use-compact-vs-basic-spacing) table can be extended with these functionalities. 
+
+* **[Expandable data table](#expandable-data-table)**
```
There is an issue with this link (I think it's because the actual heading for that section is "Expandable table" and not "Expandable **data** table"!) So should be #expandable-table !

mcoker replied:
You have "compact" and "basic" examples under this heading - should those be at the top of this list?

mmenestr commented on packages/v4/src/content/design-guidelines/usage-and-behavior/lists-and-tables/lists-and-tables.md:
```diff
@@ -70,41 +76,46 @@ The expandable list adds an expansion panel to every row to reveal more details
+* **[Compound expandable table](#compound-expandable-table)**
+* **[Sortable table](#sorting-by-columns)**
+
+### Compact table
```
Just noticed that for data list you only have a section for "compact data list" and not for "basic data list", but for tables, you have a section or both. We should definitely do one or the other for both cases... not sure which way we should go though. 

Also, I notice that for the "when to use a compact table" section, you talk about it more as "when to use compact vs basic"
but for the "when to use a basic table", you talk about it more as "when to use a table vs data list", so that's a little confusing. 

This is ones hard... @mcarrano do you think we should have seperate sections for Compact and Basic? Or should we combine into one. Also for the "when to use sections" should they focus on "table vs data list", or "compact vs basic"? (ie: for the compact table when to use, would you suggest talking about when to use a compact table vs a basic table, or would you suggest talking about when to use a table vs a data list". Keeping in mind that there is already a section later on where compact vs basic is explained....

mcarrano replied:
Wow. This is a tough question @mmenestr .  I think I see where you are headed with this. How about this idea. The sections that address List and Table usage should focus on the feature options for a list or table (not styling). Therefore use a consistent styling for all the examples. The the section called "when to use a compact table" can just focus on explaining the styling options and when to use one of the other. Does that make sense @mmenestr @doruskova ? Or am I making it more complicated?

doruskova replied:
@mmenestr @mcarrano I've added a section for a basic data list. And I agree that there's an inconsistency between these two sections. What about to explain it from both perspectives? Or merge the Compact and Basic table/data list sections? Because the compact and basic deliver the same message from a styling perspective. But, I think it's for a deeper discussion. However, I fixed everything. This is the only thing that's missing.

mmenestr commented on packages/v4/src/content/design-guidelines/usage-and-behavior/lists-and-tables/lists-and-tables.md:
```diff
@@ -145,12 +151,30 @@ If a table contains these two attributes: (System Name | Last Sync) you may want
 
```
Small thing, but Abi had recommended we stop using "ie" and start using "in other words" instead.
Same thing with "eg", she recommended we use "For example" instead.
I think it was just incase people didn't know what eg or ie meant! I don't think you've used it anywhere else but I would change it here!

doruskova replied:
Good point. It's there from the previous/current version of doc for this. I'll fix it.

mmenestr commented on packages/v4/src/content/design-guidelines/usage-and-behavior/lists-and-tables/lists-and-tables.md:
```diff
@@ -145,12 +151,30 @@ If a table contains these two attributes: (System Name | Last Sync) you may want
 
 If a table contains these three attributes: (System Name | Last Sync | Severity) you may want to show the system with the highest Severity because that is the system the user should tend to first.
 
+## When to use compact vs. basic spacing
```
This section is great! Would maybe add a little intro sentence here saying something like:
"Whether to use a compact table/data list or basic table/data list is up to you and your use case. However, here is some guidance for when to use which:"

gdoyle1 replied:
I also wonder if this should be compact vs. large spacing? I've been using the word "large" where you are using the word "basic" to describe the larger spacing variation

doruskova replied:
The "basic" word comes from the [implementation page](https://www.patternfly.org/v4/documentation/react/components/table)
But, I don't have a problem to rename it. My only concern is that word "larger" isn't that descriptive.

mcoker commented on packages/v4/src/content/design-guidelines/usage-and-behavior/lists-and-tables/lists-and-tables.md:
```diff
@@ -8,20 +8,33 @@ showTitle: true
+Think of each row in a data list as a container for some formatted content. In PatternFly 4, data list rows can accept any [layout](/documentation/react/layouts/bullseye) supported by the design system as long as all rows apply the same layout. Here are some common layouts that may be useful in a data list:
+
+* **[Grid](/documentation/react/layouts/grid):** When you want to display content in a responsive grid.
+* **[Level](/documentation/react/layouts/level):** When you want to justify content evenly over the width of the row.
```
> justify content evenly

Is that a common way of referring to what the level layout does? I would think it would say something like "vertically align content".

mcarrano replied:
I probably wrote this originally. That's what I thought the Level layout did @mcoker . But if I'm wrong, we can change this. What would you say to describe the Level?

mcoker replied:
Oh nevermind! I forgot it applies "space between" justification of items, while also centering things vertically. If "justify content evenly" conveys that, that's fine. It might also be worth noting that it also vertically aligns items in the center of the row.

FWIW, it's the `space-between` value listed here, if that provides more context - https://developer.mozilla.org/en-US/docs/Web/CSS/justify-content#Values. It applies evenly distributed spacing between items in a row.

> The items are evenly distributed within the alignment container along the main axis. The spacing between each pair of adjacent items is the same. The first item is flush with the main-start edge, and the last item is flush with the main-end edge.

mcoker commented on packages/v4/src/content/design-guidelines/usage-and-behavior/lists-and-tables/lists-and-tables.md:
```diff
@@ -8,20 +8,33 @@ showTitle: true
+
+* **[Grid](/documentation/react/layouts/grid):** When you want to display content in a responsive grid.
+* **[Level](/documentation/react/layouts/level):** When you want to justify content evenly over the width of the row.
+* **[Split](/documentation/react/layouts/split):** When you want to distribute content evenly with a main content area in the center.
```
Is there any value in also including the flex layout? That wouldn't have a specific layout, it's more of a general layout builder tool, allowing you to create a custom layout.

mcarrano replied:
Sure. Any idea about how to describe this?

mcoker replied:
Maybe "When you want a custom layout and more control over the alignment and spacing provided in the other layouts."?

mcoker commented on packages/v4/src/content/design-guidelines/usage-and-behavior/lists-and-tables/lists-and-tables.md:
```diff
@@ -8,20 +8,33 @@ showTitle: true
+* **[Split](/documentation/react/layouts/split):** When you want to distribute content evenly with a main content area in the center.
+
+### Data list capabilities
+Every [compact](#when-to-use-compact-vs-basic-spacing) or [basic](#when-to-use-compact-vs-basic-spacing) data list can be extended with these functionalities.
```
Since these link to the same place, would it just be a single link? Also "compact or basic data list" as link text would be better for a11y

doruskova replied:
Agree, makes more sense. :)

mcoker commented on packages/v4/src/content/design-guidelines/usage-and-behavior/lists-and-tables/lists-and-tables.md:
```diff
@@ -60,7 +66,7 @@ Do not use actionable data lists when actions are restricted to a single row or
+<img src="./img/expandable-list.png"  alt="Expandable data list"  width="1174"/> 
 
 
 1. **Expansion caret:** Toggles the expansion open and closed
```
```suggestion
1. **Expansion caret:** Toggles the expansion open and closed.
```

mcoker commented on packages/v4/src/content/design-guidelines/usage-and-behavior/lists-and-tables/lists-and-tables.md:
```diff
@@ -70,41 +76,46 @@ The expandable list adds an expansion panel to every row to reveal more details
 
-![standard table](./img/standard-data-table.png)
+### Table capabilities
+Every [compact](#when-to-use-compact-vs-basic-spacing) or [basic](#when-to-use-compact-vs-basic-spacing) table can be extended with these functionalities. 
```
Same here with the compact/basic links. Could it be a single link?

mcoker commented on packages/v4/src/content/design-guidelines/usage-and-behavior/lists-and-tables/lists-and-tables.md:
```diff
@@ -8,20 +8,33 @@ showTitle: true
+Every [compact](#when-to-use-compact-vs-basic-spacing) or [basic](#when-to-use-compact-vs-basic-spacing) data list can be extended with these functionalities.
+
+* **[Actionable data list](#actionable-data-list)**
+* **[Expandable data list](#expandable-data-list)**
```
Also you have compact, actionable, and expandable under this heading - should we have an entry here for `[Compact data list](#compact-data-list)`?

mcoker commented on packages/v4/src/content/design-guidelines/usage-and-behavior/lists-and-tables/lists-and-tables.md:
```diff
@@ -70,41 +76,46 @@ The expandable list adds an expansion panel to every row to reveal more details
 
-![compact table](./img/compact-data-table.png)
+<img src="./img/basic-data-table.png"  alt="Basic data table"  width="1241"/> 
 
```
There is a link in the compact table data with the text "Sorting by columns" - I think "Sorting" should be lower-case since it's in the middle of a sentence.

mcoker commented on packages/v4/src/content/design-guidelines/usage-and-behavior/lists-and-tables/lists-and-tables.md:
```diff
@@ -114,27 +125,22 @@ Use when you have more information than will comfortably fit inside a row or you
-1. **Sorted column:** When a column is being sorted by, the column header will turn blue and the sort icon will represent the direction of the sort. Subsequent clicks on the sortable column header will toggle the direction of the sort.
-2. **Hovered sort:** When a column is sortable, the sort icon will appear to the right of the column header. Upon hover, the  icon will change to a darker grey indicating that the icon is actionable.
-3. **Sortable column:** When a column is sortable, the sort icon will appear to the right of the column header in a light grey color. Sorting will not become active until the user selects the column header. This triggers the arrow to point upwards and the content to be sorted in ascending order.
+1. **Sorted column:** When a column is being sorted by, the column header will turn [blue](/design-guidelines/styles/colors#typography-and-iconography-colors) and the sort icon will represent the direction of the sort. Subsequent clicks on the sortable column header will toggle the direction of the sort.
```
Just a thought, but it seems like these should be in the order of sortable, hovered, sorted.

mcoker commented on packages/v4/src/content/design-guidelines/usage-and-behavior/lists-and-tables/lists-and-tables.md:
```diff
@@ -145,12 +151,30 @@ If a table contains these two attributes: (System Name | Last Sync) you may want
 
+## When to use compact vs. basic spacing
+
+<img src="./img/compact-vs-basic-spacing.png"  alt="Compact vs. basic spacing"  width="1237"/> 
```
I wonder if this should be 2 images, with markdown headings (that are linkable) for both "compact" and "basic". That would match the rest of the page, and make the "compact" and "basic" links I've left comments about make a little more sense if you wanted to link them individually.

doruskova replied:
That's a great idea! Thanks.

mcoker commented on packages/v4/src/content/design-guidelines/usage-and-behavior/lists-and-tables/lists-and-tables.md:
```diff
@@ -145,12 +151,30 @@ If a table contains these two attributes: (System Name | Last Sync) you may want
+<img src="./img/Mobile-data-table-example.png"  alt="Mobile data table"  width="375"/> 
+
 ## Using lists and tables in a page
 Lists or a tables should be placed in the body of a page. The width should be set by the containing element.
```
```suggestion
Lists and tables should be placed in the body of a page. The width should be set by the containing element.
```

doruskova commented:
@mmenestr @gdoyle1 @mcarrano @mcoker I've fixed everything that was mentioned here. Thanks everyone for your comments. I think that the only thing that is missing to agree on the approach of how to describe the Basic/Compact table/data list and if we should call it Basic or Larger. And the last thing is the example in the "Sorting by columns" section. I'll add some image that shows what's being described there.

mcarrano commented on packages/v4/src/content/design-guidelines/usage-and-behavior/lists-and-tables/lists-and-tables.md:
```diff
@@ -8,149 +8,201 @@ showTitle: true
 1. **[Toolbar](/design-guidelines/usage-and-behavior/toolbar):** Sits above the table and contains controls for manipulating table data. Common actions include filtering, sorting, and pagination.
-2. **Select all:** When present, selects all items in a table. If pagination is being used, this will only select items on the current page.
-3. **Column headers:** Should align with the content they contain. If the user is able to sort on a column, the first click on the header will sort the content of the table on the content in that column. Subsequent clicks will toggle the direction of the sort. Table data can only be sorted on one column at a time. See [Sorting by columns](#sorting-by-columns) for more information on the sort comp ...
+2. **[Select all](/design-guidelines/usage-and-behavior/bulk-selection):** When present, selects all items in a table. If pagination is being used, this will only select items on the current page.
```
In the library we are now calling the "Bulk selection." But I wonder if should even talk about this here as we now have a separate document that describes the bulk selection behavior https://www.patternfly.org/v4/design-guidelines/usage-and-behavior/bulk-selection Maybe reference that? Should the Data list section also reference this, too (and include in mock)?

mcarrano commented on packages/v4/src/content/design-guidelines/usage-and-behavior/lists-and-tables/lists-and-tables.md:
```diff
@@ -8,149 +8,201 @@ showTitle: true
+
+<img src="./img/compact-and-default-table.png"  alt="Compact and default table"  width="857"/> 
+
+### Table elements
```
In general we call out more things here than for the Data list. Not sure which approach is better, but we should be consistent.

mcarrano commented on packages/v4/src/content/design-guidelines/usage-and-behavior/lists-and-tables/lists-and-tables.md:
```diff
@@ -8,149 +8,201 @@ showTitle: true
-
-#### When to use
-Use a compact table when you want to show as much data per page as possible and readability is a secondary concern.
+### Table capabilities
```
What about an actionable table?

doruskova replied:
The section about the actionable table has been added.

mcarrano commented on packages/v4/src/content/design-guidelines/usage-and-behavior/lists-and-tables/lists-and-tables.md:
```diff
@@ -8,149 +8,201 @@ showTitle: true
 
 If a table contains these three attributes: (System Name | Last Sync | Severity) you may want to show the system with the highest Severity because that is the system the user should tend to first.
 
+## When to use compact vs. default spacing
```
I think that the text and recommendations are fine. But I wonder is the examples should better illustrate these differences. Right now the contents of the two tables is identical so it does not give me a good sense that one option is better that the other for the data I'm trying to display. Maybe create an example the calls for dense data (compact) and more rich graphical data that might better use default spacing?

doruskova replied:
Yeah, that makes sense. I'll create "Example" section with this approach. Thanks for this point.

doruskova commented:
@mcarrano I've updated the documentation. Let me know if there's something I can fix.

- Table and data lists have the same number of elements to keep consistency between them. 
- Changed examples for compact and default spacing + added description of examples.
- Changed the last picture where the table is placed in a card now. 
- Added section for an actionable table.
- Updated images. Can you @gdoyle1 check it out if it's ok? Thank you!

mmenestr commented:
Looks good!! My only comment is that there seems to be 3 times where you say "when to use data list vs table" or vice versa. 
1. At the very beginning, and then again underneath the table section, and the data list section. Wondering if it would be better to  consolidate these into a single section (either at the beginning, or at the end near "when to use compact vs default"). What do you think @mcarrano ?

mcarrano commented:
@mmenestr while  this is redundant, let's leave as is. Thinking forward, we may need to break this apart in the future if/when we consolidate design guidelines with component docs so it will be easier this way.

mmenestr commented:
Just gotta update images to reflect blue line on expanded rows!

doruskova commented:
@gdoyle1 @mmenestr Images have been updated. Please check, if it's correct.

mmenestr commented:
@doruskova hey, it seems like a few of the images are still missing the blue line! It should look like this for expandable tables: https://www.patternfly.org/v4/documentation/core/components/datalist#expandable

doruskova commented:
@mmenestr Yeah, sorry about that. It should be ok now. I was waiting for a library update and checkout with Gina.

mmenestr commented:
@doruskova ok just looked again and the line is right, but the shade of blue is off? If you look at the screenshot below, the lines you added look lighter blue than they should! The one below is correct. it should be #0066CC

![image](https://user-images.githubusercontent.com/51165119/88574660-a7b60700-d010-11ea-9b77-80ea8f79f80c.png)

doruskova commented:
@mmenestr The blue line has been changed to `#0066CC`

doruskova commented:
@mmenestr Hope that the images are ok now. I've added information about coloring into the "When to use" section for an expandable table.

mmenestr commented:
@doruskova the pictures LGTM now! I would just make it its own section called "Expanded panel coloring" or something along those lines, and I just tweaked the text a little bit to be more clear like this. Let me know if you don't think that's very clear though ahaha

**Expanded panel coloring**

Use a dark blue line for expanded rows:
- In a non-selectable table
- In a selectable table for a selected row. The selected row would use a dark blue line, while the other expanded un-selected rows would use a light blue line.

Use a light blue line for expanded rows:
- In a selectable table for unselected expanded rows. The unselected expanded rows would use a light blue line, while the selected expanded rows would use a dark blue line.

doruskova commented:
> @doruskova the pictures LGTM now! I would just make it its own section called "Expanded panel coloring" or something along those lines, and I just tweaked the text a little bit to be more clear like this. Let me know if you don't think that's very clear though ahaha
> 
> > **Expanded panel coloring**
> > Use a dark blue line for expanded rows:
> > 
> > * In a non-selectable table
> > * In a selectable table for a selected row. The selected row would use a dark blue line, while the other expanded un-selected rows would use a light blue line.
> > 
> > Use a light blue line for expanded rows:
> > 
> > * In a selectable table for unselected expanded rows. The unselected expanded rows would use a light blue line, while the selected expanded rows would use a dark blue line.

It makes sense to me, thank you! I've mentioned this section in a bullet point "Expansion panel" below the image.

mmenestr commented:
@doruskova just did one last thorough check, and found some things to fix (I promise this is the last time you'll have to make changes haha)
1. Many of the "bulk select" links don't work! They're all in places where you link to it twice, but the first one works and the second one doesn't
It happes under Data list elements, under actionable data list, under table elements and under actionable table!
![image](https://user-images.githubusercontent.com/51165119/88977186-6da46980-d28b-11ea-833d-14e5f730e897.png)

2. The "example" link in these sections don't work/i'm also not sure what they should be linking to?:
>**Compact spacing**
A table or data list may sometimes need to be compact to make more rows visible at a time. The more rows you can see, the less you need to use pagination. Compact spacing is recommended for data with a simple structure. See an _example_ below.

>**Default spacing**
A table or data list may sometimes need more space for rich graphical data. See an _example_ below.
 
After you fixed that it'll be good to go for real!!

doruskova commented:
@mmenestr I removed the link from the "See an example below" (they didn't make sense). But, the links about _Bulk selection, Grid, Level, Split, Flex, card views, pagination, toolbar, colors_  are external links with a relative path, so they'll work after merge. So, the links need to look like in the image below and that's why they don't work in Visual Studio Code or GitHub. 
<img width="552" alt="links" src="https://user-images.githubusercontent.com/32618096/89019550-be64a280-d31d-11ea-8d2f-be67ac3751d0.png">

When you display the rich diff in "Files changed" section on a GitHub, you can see that it generates their own links and that's why you can't see the content. We'd have to use absolute paths to see it, but you don't want to do that. Relative paths are much better. 

<img width="1331" alt="github" src="https://user-images.githubusercontent.com/32618096/89019973-75611e00-d31e-11ea-83d2-945a05d01e4f.png">

Ask someone else to confirm it and let me know :)

mmenestr commented:
@doruskova Idk what changed, but they actually all work now haha so i'm going to merge. Thank you!!

patternfly-build commented:
Your changes have been released in: 
 - patternfly-org-4@4.1.21

Thanks for your contribution! :tada:
